testworld/signature_test.go

```diff
@@ -17,12 +18,82 @@ import (
 	"github.com/centrifuge/go-centrifuge/utils"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
 )
 
+func TestHost_SignKeyNotInCollaboration(t *testing.T) {
+	// Hosts
+	alice := doctorFord.getHostTestSuite(t, "Alice")
+	mallory := doctorFord.getHostTestSuite(t, "Mallory")
+
+	actxh := testingconfig.CreateAccountContext(t, alice.host.config)
+	mctxh := testingconfig.CreateAccountContext(t, mallory.host.config)
+
+	publicKey, privateKey := GetSigningKeyPair(t, alice.host.idService, alice.id, actxh)
+
+	collaborators := [][]byte{mallory.id[:]}
+	dm := createCDWithEmbeddedPO(t, collaborators, alice.id, publicKey, privateKey, alice.host.config.GetContractAddress(config.AnchorRepo))
+
+	sr, err := dm.CalculateSigningRoot()
+	assert.NoError(t, err)
+
+	publicKeyValid, privateKeyValid := GetSigningKeyPair(t, mallory.host.idService, mallory.id, mctxh)
+	s, err := crypto.SignMessage(privateKeyValid, sr, crypto.CurveSecp256K1)
+	assert.NoError(t, err)
+
+	sig := &coredocumentpb.Signature{
+		SignatureId: append(mallory.id[:], publicKeyValid...),
+		SignerId:    mallory.id[:],
+		PublicKey:   publicKeyValid,
+		Signature:   s,
+	}
+
+	malloryDocMockSrv := mallory.host.bootstrappedCtx[documents.BootstrappedDocumentService].(*mockdoc.MockService)
+
+	malloryDocMockSrv.On("RequestDocumentSignature", mock.Anything, mock.Anything, mock.Anything).Return(sig, nil).Once()
+
+	malloryDocMockSrv.On("DeriveFromCoreDocument", mock.Anything).Return(dm, nil).Once()
+
+	//Signature verification should success
+	signatures, signatureErrors, err := alice.host.p2pClient.GetSignaturesForDocument(actxh, dm)
+
+	assert.NoError(t, err)
+	assert.Nil(t, signatureErrors)
+	fmt.Println(signatureErrors)
+	fmt.Println("-------------------------------")
+	assert.Equal(t, 1, len(signatures))
+
+	//Following simulate attack by Mallory with random keys pair
+	//Random keys pairs should cause signature verification failure
+	publicKey2, privateKey2 := GetRandomSigningKeyPair(t)
+	s, err = crypto.SignMessage(privateKey2, sr, crypto.CurveSecp256K1)
+	assert.NoError(t, err)
+
+	sig = &coredocumentpb.Signature{
+		SignatureId: append(mallory.id[:], publicKey2...),
+		SignerId:    mallory.id[:],
+		PublicKey:   publicKey2,
+		Signature:   s,
+	}
+
+	// when got request on signature of document, mocking documents.Service of Mallory return a random signature
+	malloryDocMockSrv.On("RequestDocumentSignature", mock.Anything, mock.Anything, mock.Anything).Return(sig, nil).Once()
+
+	malloryDocMockSrv.On("DeriveFromCoreDocument", mock.Anything).Return(dm, nil).Once()
+
+	//TODO
+	signatures, signatureErrors, err = alice.host.p2pClient.GetSignaturesForDocument(actxh, dm)
+	//seems to me, following should get signature verification errors but it is not.  Currenly p2p/client.go just do validateSignatureResp verification (very simple DID verification?), is this the right behavior?
```

Comment:
still failure after merging latest change, will do some investigation

Reply:
@mikiquantum  I think the root cause is this loop: https://github.com/centrifuge/go-centrifuge/blob/31427c571d89b68a9941a18fb478502e83830f7f/p2p/client.go#L336
In our updated verification logic test, due to verification failure, a null signature will be returned, but this loop ignore this error and keep doing document Anchor (which in this case will not fail because of null signature compared against failure(due to revoked key signing) of doing document Anchor before change) 
Should we keep behavior of this loop (so change the test) or  change this loop and make it return error if any returned signature verification happen error (this may lead to DOS ?)

Reply:
At the moment, it is intended to continue anchoring even if gathering signatures fail. This might change in the near future, though. The test above in travis pass and as you are asserting that there is a signature error for that collaborator, it is enough to cover the test case. 
For the `TestHost_RevokedSigningKey` test, as the signature validation now fails earlier, for now it is enough that we catch the error (as it is) but let's remove the `assert.Contains(t, message, "failed to validate signatures")` for now.